config/.env.sample

```diff
@@ -18,6 +22,13 @@ LTI_TOKEN_ENDING=/login/oauth2/token
 # Canvas JWKS endpoint (used with the LTI PLATFORM URL)
 LTI_KEYSET_ENDING=/api/lti/security/jwks
 
+# Base URL to use for auth and API calls
+CANVAS_INSTANCE_URL=
```

Comment:
Can we get this URL from launch info?

The problem we might need to address is we might state in our server configs it is test and prod. The problem we have to solve  when monthly sync happens and prod data/tools/app added to Test/beta then the LTI key would be Prod pointing to prod URL but the tool is actually launched from Test canvas instance and we might end up making changes in Prod from test env. 

This might an issue and needs a good approach to it.

Reply:
I don't think this is available in the launch info, but I can check. I don't think this is a big deal to configure, but maybe I'm not thinking about some important details. It would be nice if this was the same URL as the LTI platform, but it is not.

Reply:
In terms of the environment sync, yes, that's a problem, but feels like a separate issue. How does the current CCM handle this?

Reply:
> How does the current CCM handle this?

It is not handling it but we are opting for a caution approach (i.e documenting not to launch CCM from test)  and I want to have a better solution with new CCM.

Reply:
So it's kind of available as the base under `platformContext.launchPresentation.return_url`. I wonder if there's a custom variable for this...

If it was available, I guess I'm still not crazy about pulling it as a launch variable and storing it in the session. It could be done, but then I'd have to pass that variable from the session on the controller function into each method call on `CanvasService`. Feels clunky; it seems reasonable to define the base URL once in the configuration alongside the API client and secret.

Reply:
> I don't think this is available in the launch info, but I can check.

`$Canvas.api.baseUrl` custom launch variable

Reply:
Okay, so it's available, now we have to decide if it's actually a good idea. I think might point(s) above are reasonable.

Reply:
> Feels clunky; it seems reasonable to define the base URL once in the configuration alongside the API client and secret.

In that case if you see launch is happening from Test and Canvas_INSTANCE_URL is Prod we should error out this case. We need to add some validation around this. What do you say?

Reply:
Let me think about this some more, do some testing.

Reply:
@pushyamig, do you like this better? https://github.com/ssciolla/canvas-course-manager-next/commit/2949f906b7a6531f0b43e3027812e4963e128d21 This is what you originally suggested. The delayed setting of `url` feels awkward to me code-wise, but I can see the benefit in terms of keeping things in sync. Alternatively, we could just compare `canvasConfig.instanceURL` with `customLTIVariables.api_base_url` in `LTIService`, throw an error if they don't match.

Reply:
Background: If the Prod LTI configuration (synced to test instance) will have Prod URL and will probably goes to Prod CCM server. But the LTI launch is happening in test, the launch info would have test URL. 

Use case 1: If we don't set the Canvas instance URL in the config and proceed with launch info data then we probably make changes to Test instance data. Like cross list a section in Test course. We is not a big issue but seems to be wrong. 
Use case 2: But the   of the above usecase is bad. i.e Test CCM launching from Prod. 

Approach1: Probably the right approach would be set the Canvas instance URL in config and also get the URL from custom params and if there is a mismatch then throw an error. We don't want to entertain this case. I don't see a value in this.

Approach 2: If you don't want to deal with setting Canvas instance URL in configuration then another thing would be since we are setting CCM Domain in config we can decide if CCM server is test or prod server and based on custom param launch URL we could reject the use case.

We can discuss offline if you like